File: PODM/SW/pod-manager/podm-business-entities/src/main/java/com/intel/podm/business/entities/redfish/MetricDefinition.java
/*
 * Copyright (c) 2017 Intel Corporation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.intel.podm.business.entities.redfish;

import com.intel.podm.business.entities.IgnoreUnlinkingRelationship;
import com.intel.podm.business.entities.redfish.base.DiscoverableEntity;
import com.intel.podm.business.entities.redfish.base.Entity;
import com.intel.podm.business.entities.redfish.embeddables.CalculationParamsType;
import com.intel.podm.common.types.Calculable;
import com.intel.podm.common.types.CalculationAlgorithm;
import com.intel.podm.common.types.DataType;
import com.intel.podm.common.types.Id;
import com.intel.podm.common.types.Implementation;
import com.intel.podm.common.types.MetricType;
import com.intel.podm.common.types.MetricValueType;
import com.intel.podm.common.types.PhysicalContext;
import com.intel.podm.common.types.SensorType;

import javax.persistence.CollectionTable;
import javax.persistence.Column;
import javax.persistence.ElementCollection;
import javax.persistence.Enumerated;
import javax.persistence.Index;
import javax.persistence.JoinColumn;
import javax.persistence.NamedQueries;
import javax.persistence.NamedQuery;
import javax.persistence.OneToMany;
import javax.persistence.OneToOne;
import javax.persistence.OrderColumn;
import javax.persistence.Table;
import java.math.BigDecimal;
import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

import static com.intel.podm.common.utils.Contracts.requiresNonNull;
import static javax.persistence.CascadeType.MERGE;
import static javax.persistence.CascadeType.PERSIST;
import static javax.persistence.CascadeType.REMOVE;
import static javax.persistence.EnumType.STRING;
import static javax.persistence.FetchType.LAZY;

@javax.persistence.Entity
@NamedQueries({
    @NamedQuery(name = MetricDefinition.GET_ALL_METRIC_DEFINITIONS_IDS,
        query = "SELECT metricDefinition.entityId FROM MetricDefinition metricDefinition")
})
@Table(name = "metric_definition", indexes = @Index(name = "idx_metric_definition_entity_id", columnList = "entity_id", unique = true))
@SuppressWarnings({"checkstyle:MethodCount", "checkstyle:ClassFanOutComplexity"})
public class MetricDefinition extends DiscoverableEntity {
    public static final String GET_ALL_METRIC_DEFINITIONS_IDS = "GET_ALL_METRIC_DEFINITIONS_IDS";

    @Column(name = "entity_id", columnDefinition = ENTITY_ID_STRING_COLUMN_DEFINITION)
    private Id entityId;

    @Column(name = "sensor_type")
    @Enumerated(STRING)
    private SensorType sensorType;

    @Column(name = "calculable")
    @Enumerated(STRING)
    private Calculable calculable;

    @Column(name = "data_type")
    @Enumerated(STRING)
    private DataType dataType;

    @Column(name = "implementation")
    @Enumerated(STRING)
    private Implementation implementation;

    @Column(name = "sensing_interval")
    private String sensingInterval;

    @Column(name = "is_linear")
    private Boolean isLinear;

    @Column(name = "metric_type")
    @Enumerated(STRING)
    private MetricType metricType;

    @Column(name = "physical_context")
    @Enumerated(STRING)
    private PhysicalContext physicalContext;

    @Column(name = "calculation_algorithm")
    @Enumerated(STRING)
    private CalculationAlgorithm calculationAlgorithm;

    @Column(name = "units")
    private String units;

    @Column(name = "min_reading_range")
    private BigDecimal minReadingRange;

    @Column(name = "max_reading_range")
    private BigDecimal maxReadingRange;

    @Column(name = "precision")
    private Integer precision;

    @Column(name = "calibration")
    private BigDecimal calibration;

    @Column(name = "accuracy")
    private BigDecimal accuracy;

    @Column(name = "time_stamp_accuracy")
    private String timeStampAccuracy;

    @Column(name = "calculation_time_interval")
    private String calculationTimeInterval;

    @Column(name = "calculation_precision")
    private Double calculationPrecision;

    @Column(name = "discrete_metric_type")
    @Enumerated(STRING)
    private MetricValueType discreteMetricType;

    @ElementCollection
    @CollectionTable(name = "metric_definition_metric_properties", joinColumns = @JoinColumn(name = "metric_definition_id"))
    @Column(name = "metric_properties")
    @OrderColumn(name = "metric_properties_order")
    private List<String> metricProperties = new ArrayList<>();

    @ElementCollection
    @CollectionTable(name = "metric_definition_discrete_values", joinColumns = @JoinColumn(name = "metric_definition_id"))
    @Column(name = "discrete_values")
    @OrderColumn(name = "discrete_values_order")
    private List<String> discreteValues = new ArrayList<>();

    @OneToOne(fetch = LAZY, cascade = {MERGE, PERSIST})
    @JoinColumn(name = "metric_item_id")
    private MetricItem metricItem;

    @IgnoreUnlinkingRelationship
    @OneToMany(fetch = LAZY, cascade = {MERGE, PERSIST, REMOVE})
    @JoinColumn(name = "metric_definition_wildcard_id")
    private List<MetricDefinitionWildcard> wildcards = new ArrayList<>();

    @ElementCollection
    @CollectionTable(name = "metric_definition_calculation_parameters", joinColumns = @JoinColumn(name = "metric_definition_id"))
    @OrderColumn(name = "calculation_parameters_values_order")
    private List<CalculationParamsType> calculationParameters = new ArrayList<>();

    @Override
    public Id getId() {
        return entityId;
    }

    @Override
    public void setId(Id id) {
        entityId = id;
    }

    public SensorType getSensorType() {
        return sensorType;
    }

    public void setSensorType(SensorType sensorType) {
        this.sensorType = sensorType;
    }

    public Implementation getImplementation() {
        return implementation;
    }

    public void setImplementation(Implementation implementation) {
        this.implementation = implementation;
    }

    public String getSensingInterval() {
        return sensingInterval;
    }

    public void setSensingInterval(String sensingInterval) {
        this.sensingInterval = sensingInterval;
    }

    public MetricType getMetricType() {
        return metricType;
    }

    public void setMetricType(MetricType metricType) {
        this.metricType = metricType;
    }

    public PhysicalContext getPhysicalContext() {
        return physicalContext;
    }

    public void setPhysicalContext(PhysicalContext physicalContext) {
        this.physicalContext = physicalContext;
    }

    public String getUnits() {
        return units;
    }

    public void setUnits(String readingUnits) {
        this.units = readingUnits;
    }

    public BigDecimal getMinReadingRange() {
        return minReadingRange;
    }

    public void setMinReadingRange(BigDecimal minReadingRange) {
        this.minReadingRange = minReadingRange;
    }

    public BigDecimal getMaxReadingRange() {
        return maxReadingRange;
    }

    public void setMaxReadingRange(BigDecimal maxReadingRange) {
        this.maxReadingRange = maxReadingRange;
    }

    public Integer getPrecision() {
        return precision;
    }

    public void setPrecision(Integer precision) {
        this.precision = precision;
    }

    public BigDecimal getCalibration() {
        return calibration;
    }

    public void setCalibration(BigDecimal calibration) {
        this.calibration = calibration;
    }

    public List<String> getMetricProperties() {
        return metricProperties;
    }

    public void addMetricProperties(String metricProperties) {
        this.metricProperties.add(metricProperties);
    }

    public Calculable getCalculable() {
        return calculable;
    }

    public void setCalculable(Calculable calculable) {
        this.calculable = calculable;
    }

    public DataType getDataType() {
        return dataType;
    }

    public void setDataType(DataType dataType) {
        this.dataType = dataType;
    }

    public Boolean getLinear() {
        return isLinear;
    }

    public void setLinear(Boolean linear) {
        isLinear = linear;
    }

    public BigDecimal getAccuracy() {
        return accuracy;
    }

    public void setAccuracy(BigDecimal accuracy) {
        this.accuracy = accuracy;
    }

    public String getTimeStampAccuracy() {
        return timeStampAccuracy;
    }

    public void setTimeStampAccuracy(String timeStampAccuracy) {
        this.timeStampAccuracy = timeStampAccuracy;
    }

    public String getCalculationTimeInterval() {
        return calculationTimeInterval;
    }

    public void setCalculationTimeInterval(String calculationTimeInterval) {
        this.calculationTimeInterval = calculationTimeInterval;
    }

    public CalculationAlgorithm getCalculationAlgorithm() {
        return calculationAlgorithm;
    }

    public void setCalculationAlgorithm(CalculationAlgorithm calculationAlgorithm) {
        this.calculationAlgorithm = calculationAlgorithm;
    }

    public Double getCalculationPrecision() {
        return calculationPrecision;
    }

    public void setCalculationPrecision(Double calculationPrecision) {
        this.calculationPrecision = calculationPrecision;
    }

    public MetricValueType getDiscreteMetricType() {
        return discreteMetricType;
    }

    public void setDiscreteMetricType(MetricValueType discreteMetricType) {
        this.discreteMetricType = discreteMetricType;
    }

    public List<String> getDiscreteValues() {
        return discreteValues;
    }

    public void addDiscreteValues(String discreteValue) {
        this.discreteValues.add(discreteValue);
    }

    public List<MetricDefinitionWildcard> getWildcards() {
        return wildcards;
    }

    public void addWildcards(MetricDefinitionWildcard wildcard) {
        requiresNonNull(wildcard, "wildcard");
        this.wildcards.add(wildcard);
    }

    public List<CalculationParamsType> getCalculationParameters() {
        return calculationParameters;
    }

    public void addCalculationParameters(CalculationParamsType calculationParameter) {
        this.calculationParameters.add(calculationParameter);
    }

    public MetricItem getMetricItem() {
        return metricItem;
    }

    public void setMetricItem(MetricItem metricItem) {
        if (!Objects.equals(this.metricItem, metricItem)) {
            unlinkMetricItem(this.metricItem);
            this.metricItem = metricItem;
            if (metricItem != null && !this.equals(metricItem.getMetricDefinition())) {
                metricItem.setMetricDefinition(this);
            }
        }
    }

    public void unlinkMetricItem(MetricItem metricItem) {
        if (Objects.equals(this.metricItem, metricItem)) {
            this.metricItem = null;
            if (metricItem != null) {
                metricItem.unlinkMetricDefinition(this);
            }
        }
    }

    @Override
    public void preRemove() {
        unlinkMetricItem(metricItem);
    }

    @Override
    public boolean containedBy(Entity possibleParent) {
        return false;
    }
}

File: PSME/common/agent-framework/src/module/model/metric_definition.cpp
/*!
 * @copyright Copyright (c) 2017-2019 Intel Corporation.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * @file metric_definition.cpp
 */


#include "agent-framework/module/model/metric_definition.hpp"
#include "agent-framework/module/constants/common.hpp"
#include "agent-framework/module/utils/iso8601_time_interval.hpp"

using namespace agent_framework::model;
using agent_framework::utils::Iso8601TimeInterval;

const enums::Component MetricDefinition::component = enums::Component::MetricDefinition;
const enums::CollectionName MetricDefinition::collection_name = enums::CollectionName::MetricDefinitions;

MetricDefinition::MetricDefinition(const std::string& parent_uuid, enums::Component parent_type) :
    Resource{parent_uuid, parent_type} { }


MetricDefinition::~MetricDefinition() { }


MetricDefinition MetricDefinition::from_json(const json::Json& json) {
    MetricDefinition metric{};

    /* properties not allowed to be set */
    metric.set_name(json[literals::MetricDefinition::NAME]);
    metric.set_data_type(json[literals::MetricDefinition::DATA_TYPE]);
    metric.set_metric_type(json[literals::MetricDefinition::METRIC_TYPE]);
    metric.set_discrete_metric_type(json[literals::MetricDefinition::DISCRETE_METRIC_TYPE]);
    metric.set_discrete_values(DiscreteValues::from_json(json[literals::MetricDefinition::DISCRETE_VALUES]));

    metric.fill_from_json(json);
    return metric;
}

void MetricDefinition::fill_from_json(const json::Json& json) {
    if (json.count(literals::MetricDefinition::DESCRIPTION) != 0) {
        set_description(json[literals::MetricDefinition::DESCRIPTION]);
    }
    if (json.count(literals::MetricDefinition::IMPLEMENTATION) != 0) {
        set_implementation(json[literals::MetricDefinition::IMPLEMENTATION]);
    }
    if (json.count(literals::MetricDefinition::CALCULABLE) != 0) {
        set_calculable(json[literals::MetricDefinition::CALCULABLE]);
    }
    if (json.count(literals::MetricDefinition::UNITS) != 0) {
        set_units(json[literals::MetricDefinition::UNITS]);
    }
    if (json.count(literals::MetricDefinition::IS_LINEAR) != 0) {
        set_is_linear(json[literals::MetricDefinition::IS_LINEAR]);
    }
    if (json.count(literals::MetricDefinition::WILDCARDS) != 0) {
        set_wildcards(Wildcards::from_json(json[literals::MetricDefinition::WILDCARDS]));
    }
    if (json.count(literals::MetricDefinition::CALCULATION_PARAMETERS) != 0) {
        set_calculation_parameters(CalculationParametersArray::from_json(
            json[literals::MetricDefinition::CALCULATION_PARAMETERS]));
    }
    if (json.count(literals::MetricDefinition::PHYSICAL_CONTEXT) != 0) {
        set_physical_context(json[literals::MetricDefinition::PHYSICAL_CONTEXT]);
    }
    if (json.count(literals::MetricDefinition::SENSOR_TYPE) != 0) {
        set_sensor_type(json[literals::MetricDefinition::SENSOR_TYPE]);
    }
    if (json.count(literals::MetricDefinition::SENSING_INTERVAL) != 0) {
        set_sensing_interval(json[literals::MetricDefinition::SENSING_INTERVAL]);
    }
    if (json.count(literals::MetricDefinition::PRECISION) != 0) {
        set_precision(json[literals::MetricDefinition::PRECISION]);
    }
    if (json.count(literals::MetricDefinition::ACCURACY) != 0) {
        set_accuracy(json[literals::MetricDefinition::ACCURACY]);
    }
    if (json.count(literals::MetricDefinition::CALIBRATION) != 0) {
        set_calibration(json[literals::MetricDefinition::CALIBRATION]);
    }
    if (json.count(literals::MetricDefinition::TIME_STAMP_ACCURACY) != 0) {
        set_time_stamp_accuracy(json[literals::MetricDefinition::TIME_STAMP_ACCURACY]);
    }
    if (json.count(literals::MetricDefinition::MIN_READING_RANGE) != 0) {
        set_min_reading_range(json[literals::MetricDefinition::MIN_READING_RANGE]);
    }
    if (json.count(literals::MetricDefinition::MAX_READING_RANGE) != 0) {
        set_max_reading_range(json[literals::MetricDefinition::MAX_READING_RANGE]);
    }
    if (json.count(literals::MetricDefinition::CALCULATION_ALGORITHM) != 0) {
        set_calculation_algorithm(json[literals::MetricDefinition::CALCULATION_ALGORITHM]);
    }
    if (json.count(literals::MetricDefinition::CALCULATION_TIME_INTERVAL) != 0) {
        set_calculation_time_interval(json[literals::MetricDefinition::CALCULATION_TIME_INTERVAL]);
    }
    if (json.count(literals::MetricDefinition::CALCULATION_PRECISION) != 0) {
        set_calculation_precision(json[literals::MetricDefinition::CALCULATION_PRECISION]);
    }

    /* special property: not in JSON, but handled via config file */
    if (json.count(literals::MetricDefinition::SHOREUP_PERIOD) != 0) {
        if (json[literals::MetricDefinition::SHOREUP_PERIOD].is_string()) {
            set_shoreup_period(json[literals::MetricDefinition::SHOREUP_PERIOD].get<std::string>());
        }
        else if (json[literals::MetricDefinition::SHOREUP_PERIOD].is_number()) {
            set_shoreup_period(json[literals::MetricDefinition::SHOREUP_PERIOD].get<double>());
        }
        else {
            log_error(LOGUSR, "Shore up interval must be either ISO8601 string or seconds as a number");
        }
    }
}


json::Json MetricDefinition::to_json() const {
    json::Json json = json::Json();
    json[literals::MetricDefinition::NAME] = get_name();
    json[literals::MetricDefinition::DESCRIPTION] = get_description();
    json[literals::MetricDefinition::IMPLEMENTATION] = get_implementation();
    json[literals::MetricDefinition::CALCULABLE] = get_calculable();
    json[literals::MetricDefinition::UNITS] = get_units();
    json[literals::MetricDefinition::DATA_TYPE] = get_data_type();
    json[literals::MetricDefinition::IS_LINEAR] = get_is_linear();
    json[literals::MetricDefinition::METRIC_TYPE] = get_metric_type();
    json[literals::MetricDefinition::WILDCARDS] = get_wildcards().to_json();
    json[literals::MetricDefinition::CALCULATION_PARAMETERS] = get_calculation_parameters().to_json();
    json[literals::MetricDefinition::PHYSICAL_CONTEXT] = get_physical_context();
    json[literals::MetricDefinition::SENSOR_TYPE] = get_sensor_type();
    json[literals::MetricDefinition::SENSING_INTERVAL] = get_sensing_interval();
    json[literals::MetricDefinition::DISCRETE_VALUES] = get_discrete_values().to_json();
    json[literals::MetricDefinition::PRECISION] = get_precision();
    json[literals::MetricDefinition::ACCURACY] = get_accuracy();
    json[literals::MetricDefinition::CALIBRATION] = get_calibration();
    json[literals::MetricDefinition::TIME_STAMP_ACCURACY] = get_time_stamp_accuracy();
    json[literals::MetricDefinition::MIN_READING_RANGE] = get_min_reading_range();
    json[literals::MetricDefinition::MAX_READING_RANGE] = get_max_reading_range();
    json[literals::MetricDefinition::CALCULATION_ALGORITHM] = get_calculation_algorithm();
    json[literals::MetricDefinition::CALCULATION_TIME_INTERVAL] = get_calculation_time_interval();
    json[literals::MetricDefinition::DISCRETE_METRIC_TYPE] = get_discrete_metric_type();
    json[literals::MetricDefinition::CALCULATION_PRECISION] = get_calculation_precision();
    return json;
}

void MetricDefinition::set_sensing_interval(const OptionalField<std::string>& sensing_interval) {
    m_sensing_interval = sensing_interval;
    if (sensing_interval.has_value()) {
        m_sensing_period = Iso8601TimeInterval::parse(*sensing_interval);
    }
}

void MetricDefinition::set_calculation_time_interval(const OptionalField<std::string>& calculation_time_interval) {
    m_calculation_time_interval = calculation_time_interval;
    if (calculation_time_interval.has_value()) {
        m_calculation_period = Iso8601TimeInterval::parse(*calculation_time_interval);
    }
}

void MetricDefinition::set_shoreup_period(const std::string& period) {
    m_shoreup_period = Iso8601TimeInterval::parse(period).as<std::chrono::duration<double>>().count();
}

void MetricDefinition::set_shoreup_period(double period) {
    m_shoreup_period = period;
}
